File: src/main/java/com/github/ramonwirsch/fopRenderer/XMLValidatorTask.java
package com.github.ramonwirsch.fopRenderer;

import com.github.ramonwirsch.fopRenderer.util.CollectionUtil;
import org.gradle.api.DefaultTask;
import org.gradle.api.Project;
import org.gradle.api.file.FileCollection;
import org.gradle.api.tasks.Input;
import org.gradle.api.tasks.InputFiles;
import org.gradle.api.tasks.Internal;
import org.gradle.api.tasks.ParallelizableTask;
import org.gradle.api.tasks.TaskAction;
import org.gradle.api.tasks.TaskExecutionException;
import org.gradle.api.tasks.TaskInstantiationException;
import org.gradle.api.tasks.incremental.IncrementalTaskInputs;
import org.slf4j.Logger;
import org.xml.sax.SAXException;

import javax.xml.parsers.ParserConfigurationException;
import java.io.File;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.HashSet;
import java.util.Set;

/**
 * Created by ramonw on 06.11.15.
 */
@ParallelizableTask
public class XMLValidatorTask extends DefaultTask {

	private final Logger logger = getLogger();
	private SchemaConfigExtension schemaConfig;
	private boolean fullValidation = false;

    public XMLValidatorTask() {
		setGroup("verification");
	}

	@InputFiles
	public FileCollection getInputFiles() {
		return schemaConfig.getFiles();
	}

	@Internal
	public SchemaConfigExtension getSchemaConfig() {
		return schemaConfig;
	}

	public void setSchemaConfig(SchemaConfigExtension schemaConfig) {
		this.schemaConfig = schemaConfig;
	}

	@InputFiles
	public FileCollection getSchemaFiles() {
		Project project = getProject();
		File dir = schemaConfig.getSchemaDir() != null ? schemaConfig.getSchemaDir() : project.getProjectDir();
		return project.fileTree(CollectionUtil.immutableMap("dir", dir, "include", "*.xsd"));
	}

    @Input
	public URL getSchemaUri() {
		if (FopRendererPlugin.isOffline(getProject()) || schemaConfig.getSchemaUri() != null) {
			if (schemaConfig.getOfflineSchema() == null)
				throw new TaskInstantiationException(String.format("Offline schema validation requested but no offline schema given for '%s'", schemaConfig.getName()));
			else
				try {
					return schemaConfig.getOfflineSchema().toURI().toURL();
				} catch (MalformedURLException e) {
					throw new RuntimeException(e);
				}
		} else if (!schemaConfig.isUseInherentSchemas())
			return schemaConfig.getSchemaUri();
		else
			return SchemaConfigExtension.FALLBACK_URL;
	}

	@Override
	@Internal
	public String getDescription() {
		return String.format("Validate all files of SchemaConfig group '%s'", schemaConfig.getName());
	}

	@Internal
	public boolean isFullValidation() {
		return fullValidation;
	}

	public void setFullValidation(boolean fullValidation) {
		this.fullValidation = fullValidation;
	}

	@TaskAction
	void execute(IncrementalTaskInputs inputs) throws ParserConfigurationException, SAXException {
		XMLValidator validator;
		if (!FopRendererPlugin.isOffline(getProject()) && schemaConfig.isUseInherentSchemas()) {
			logger.info("Using inherent schemas");
			validator = new XMLValidator(null);
		} else {
			URL schemaUri = getSchemaUri();
			logger.info("Using schema {}", schemaUri);
			validator = new XMLValidator(schemaUri);
		}

        boolean success = true;

		final Set<File> toValidate = new HashSet<>();

		if (!fullValidation) {
			inputs.outOfDate(change -> {
				if (fullValidation)
					return;

				File file = change.getFile();
				String name = file.getName();

				if (name.endsWith(".xsd")) {
					fullValidation = true;
					return;
				}

				logger.info("out of date: {}", name);
				toValidate.add(file);
			});

			inputs.removed(change -> {
				String name = change.getFile().getName();
				if (name.endsWith(".xsd")) {
					fullValidation = true;
					return;
				}

				logger.info("removed: {}", name);
			});
		}

		if (fullValidation) {
			toValidate.addAll(getSchemaConfig().getFiles().getFiles());
			logger.warn("schema change detected, full validation required!");
		}

		success = toValidate.stream().allMatch(validator::validate);

        if (!success)
			throw new TaskExecutionException(this, new Exception("One of the inputs failed to validate. See log output"));
	}
}

File: src/main/kotlin/com/github/ramonwirsch/fopRenderer/validation/XMLValidatorTask.kt
package com.github.ramonwirsch.fopRenderer.validation

import com.github.ramonwirsch.fopRenderer.FopRendererPlugin
import com.github.ramonwirsch.fopRenderer.SchemaConfigExtension
import org.gradle.api.DefaultTask
import org.gradle.api.file.FileCollection
import org.gradle.api.file.FileType
import org.gradle.api.file.RegularFileProperty
import org.gradle.api.provider.Property
import org.gradle.api.tasks.*
import org.gradle.work.ChangeType
import org.gradle.work.Incremental
import org.gradle.work.InputChanges
import org.gradle.workers.WorkAction
import org.gradle.workers.WorkParameters
import org.gradle.workers.WorkerExecutor
import org.xml.sax.SAXException
import java.io.File
import java.net.MalformedURLException
import java.net.URL
import java.util.*
import javax.inject.Inject
import javax.xml.parsers.ParserConfigurationException

open class XMLValidatorTask
@Inject constructor(
		private val workerExecutor: WorkerExecutor
) : DefaultTask() {

	@get:Internal
	lateinit var schemaConfig: SchemaConfigExtension
	@get:Internal
	var isFullValidation = false

	init {
		group = "verification"
		outputs.upToDateWhen { true }
	}

	val inputFiles: FileCollection
		@Incremental
		@InputFiles
		@PathSensitive(PathSensitivity.RELATIVE)
		get() = schemaConfig.files

	val schemaFiles: FileCollection
		@InputFiles
		@PathSensitive(PathSensitivity.RELATIVE)
		get() {
			val project = project
			val dir = schemaConfig.schemaDir ?: project.projectDir
			return project.fileTree(dir) {
				include("*.xsd")
			}
		}

	val schemaUri: URL
		@Input
		get() {
			if (FopRendererPlugin.isOffline(project) || schemaConfig.schemaUri == null) {
				val offlineSchema = schemaConfig.offlineSchema
				if (offlineSchema == null)
					throw TaskInstantiationException(String.format("Offline schema validation requested but no offline schema given for '%s'", schemaConfig.name))
				else
					try {
						return offlineSchema.toURI().toURL()
					} catch (e: MalformedURLException) {
						throw RuntimeException(e)
					}

			} else return if (!schemaConfig.isUseInherentSchemas)
				schemaConfig.schemaUri!!
			else
				SchemaConfigExtension.FALLBACK_URL
		}

	@Internal
	override fun getDescription() = "Validate all files of SchemaConfig group '${schemaConfig.name}'"

	@TaskAction
	@Throws(ParserConfigurationException::class, SAXException::class)
	internal fun execute(inputs: InputChanges) {
		val validationSchema = if (!FopRendererPlugin.isOffline(project) && schemaConfig.isUseInherentSchemas) {
			logger.info("Using inherent schemas")
			SchemaConfigExtension.FALLBACK_URL
		} else {
			val schemaUri = schemaUri
			logger.info("Using schema {}", schemaUri)
			schemaUri
		}

		val toValidate = HashSet<File>()

		if (!isFullValidation && inputs.isIncremental) {
			inputs.getFileChanges(inputFiles).forEach { change ->
				when {
					change.fileType == FileType.DIRECTORY -> {
						// Nothing to do, still up-to-date
					}
					change.changeType == ChangeType.REMOVED -> {
						logger.info("removed: {}", change.file)
						// Nothing To do, still up-to-date
					}
					else -> {
						val file = change.file
						logger.info("XML out of date: {}", file)
						toValidate += file
					}
				}
			}
		}

		if (isFullValidation || !inputs.isIncremental) {
			toValidate.addAll(schemaConfig.files.files)
			logger.warn("Validating all inputs")
		}

		toValidate.forEach { v ->
			workerExecutor.noIsolation().submit(ValidationWorker::class.java) {
				this.validationSchema.set(validationSchema)
				input.set(v)
			}
		}
	}

	interface ValidationParameters: WorkParameters {
		val validationSchema: Property<URL>
		val input: RegularFileProperty
	}

	abstract class ValidationWorker: WorkAction<ValidationParameters> {

		override fun execute() {
			val params = parameters
			val validationSchema = params.validationSchema.get()

			val validator = XMLValidatorFactory.forSchema(validationSchema).createValidator()

			validator.validateOrThrow(params.input.asFile.get())
		}
	}
}